### DISCUSSION TITLE
GET method doesn't support body payload

### DISCUSSION THREAD

**Discussion by ZhymabekRoman** [encode/httpx#2277] [Q&A] [answered]
It would be nice to be able to send a body with a GET request. I understand that this may not be considered a good practice, but this is necessary for the API that I have to work with.

RFC 2616, section 4.3 clearly states:

> A message-body **MUST NOT be included** in a request **if the specification of the request method** (section 5.1.1) **does not allow sending an entity-body** in requests.
> https://tools.ietf.org/html/rfc2616#section-4.3

_However_, in the entirety of [section 9.3](https://tools.ietf.org/html/rfc2616#section-9.3), the section defining the GET verb, **nothing prevents a GET request from having a body**.

But before you draw any conclusions - the functionality defined for the GET verb also **does not include any logic involving the message body**.

In other words, if we are to follow the specification:

1. **It is possible** to send a message body with a GET request per specification.
2. However, the server responding to the GET request **must ignore the body** to follow the standard.

Essentially, there is no point, per standard, to send a body with a GET request, even though it is not explicitly disallowed.

[Roy T. Fielding](http://roy.gbiv.com/) backs this interpretation up:

> ...**any HTTP request message is allowed to contain a message body**, and thus must parse messages with that in mind. **Server semantics for GET**, however, are restricted such that **a body**, if any, **has no semantic meaning to the request**. The requirements on parsing are separate from the requirements on method semantics.
> So, yes, **you can send a body with GET, and no, it is never useful to do so**.
> https://groups.yahoo.com/neo/groups/rest-discuss/conversations/messages/9962

More about this: https://github.com/swagger-api/swagger-ui/issues/2136

**Comment by ZhymabekRoman** [comment]
I know there is a way around this using `httpx.request("GET", data=...)`, but this is a workaround, in requests this functionality is implemented out of the box

**Comment by florimondmanca** [comment, marked as answer]
Hi @ZhymabekRoman 

HTTPX does not _disallow_ sending bodies on any method. We don’t allow it in the shortcut methods, but it is fully allowed with the usage you mentioned, which is not a workaround but the documented and supported way to do so, as per the Requests compatibility guide:

https://www.python-httpx.org/compatibility/#request-body-on-http-methods

> If you really do need to send request data using these http methods you should use the generic .request function instead.

```
httpx.request(
  method="DELETE",
  url="https://www.example.com/",
  content=b'A request body on a DELETE request.'
)
```

There has been extensive thinking behind this design decision and various discussions already. Please feel free to search through the issue tracker.

I’ll move this to a discussion. Thanks!
